obspy/core/event/resourceid.py

```diff
@@ -0,0 +1,653 @@
+        or the bound object has been garbage collected, try to find another
+        object that was assigned the same resource_id.
+
+        The parent_id_tree will be scanned first to see if any object_ids
+        have been assigned to the same resource_id in the parent scope. If
+        not simply use the last object assigned the resource_id. If no object
+        is found return None.
+        """
+        # Warn if resource_id was bound but its object got gc'ed.
+        if self._object_id is not None:
+            msg = ("The object with identity of: %d no longer exists, "
+                   "returning the most recently created object with a"
+                   " resource id of: %s") % (self._object_id, self.id)
+            warnings.warn(msg, UserWarning)
+        # try to get the object_id from the parent_id_tree
+        id_tree = ResourceIdentifier._parent_id_tree
+        try:
+            obj_id = id_tree[self._parent_key][self._resource_key]
+        except (KeyError, TypeError):
+            pass
+        else:
+            out = ResourceIdentifier._id_object_map[obj_id]
+            self.set_referred_object(out, warn=False)
+            return out
+        # else try to find the last object assigned the same resource_id
+        try:
+            rid_list = ResourceIdentifier._id_order[self._resource_key]
+        except KeyError:  # no existing object is assigned to this resource_id
+            return None
+        while len(rid_list):
+            try:
+                obj_id = ResourceIdentifier._id_order[self._resource_key][-1]
+            except KeyError:
+                return None
+            else:
+                try:
+                    obj = ResourceIdentifier._id_object_map[obj_id]
+                    self.set_referred_object(obj, warn=False)
+                    return obj
+                except KeyError:  # object got gc'ed or was not defined
+                    rid_list.pop()  # remove last element in list
+
+    def set_referred_object(self, referred_object, warn=True, parent=None):
+        """
+        Bind an object to the ResourceIdentifier instance.
+
+        Also allows the object to be bound to a parent scope.
+
+        :param referred_object: The object to which the resource id refers.
+        :type referred_object: object
+        :param warn:
+            If True, issue a warning if the referred_object is not equal to
+            the last object assigned the same resource_id.
+        :type warn: bool
+        :param parent:
+            An object or int (id) to which the resource_id should be scoped.
+            Used, for example, to ensure the all resource ids belonging to an
+            event object are event-scoped.
+        :type parent: object, int
+        """
+        self._object_id = id(referred_object)
+        # Use parent-scoping feature if parent is not None.
+        if parent is not None:
+            self._parent_key = parent
```

Comment:
What would happen if this is called multiple times with different parent objects?

Reply:
I am trying to remember (I had this really well thought out at one point) but I think this does happen when an event is copied. I don't think this will cause issues since the resource_id instance will now point to the correct object in the new scope.